File: secure-code-review/vulnerable-examples/javascript/injection/sql_injection.js
/**
 * SQL Injection — vulnerable and secure variants, side by side.
 *
 * Scenario: an authenticated "search my orders" endpoint in a Node.js
 * service using `mysql2`. The endpoint filters the current user's
 * orders by an optional `customer_name` query parameter.
 *
 * The vulnerable version uses a template literal to build the SQL —
 * the pattern that continues to ship to production because "just this
 * one value" feels safer than it is. The secure version uses
 * placeholder-based parameterization, which is what mysql2's
 * `execute()` provides.
 *
 * Run against a throwaway database; the demo at the bottom shows the
 * vulnerable handler leaking every row while the secure handler
 * returns zero rows for the same attacker payload.
 */

const mysql = require("mysql2/promise");

// ---------------------------------------------------------------------------
// VULNERABLE: template-literal interpolation into a raw query.
// ---------------------------------------------------------------------------
async function searchOrdersVulnerable(conn, userId, customerName) {
  // Exploit payloads (sent as the `customer_name` query parameter):
  //
  //   "' OR '1'='1"
  //      → WHERE user_id = 42 AND customer_name = '' OR '1'='1'
  //      → returns every row for every user — tenant isolation broken.
  //
  //   "' UNION SELECT id, email, password_hash FROM users --"
  //      → returns the users table through the orders endpoint.
  //
  //   "'; DROP TABLE orders; --"
  //      → if multi-statement is enabled on the connection, destructive.
  //
  // Bonus: even without single-quote injection, a numeric field
  // would be exploitable the same way with no quotes at all.
  const sql = `
    SELECT id, customer_name, total_cents, created_at
      FROM orders
     WHERE user_id = ${userId}
       AND customer_name = '${customerName}'
  `;
  const [rows] = await conn.query(sql);
  return rows;
}

// ---------------------------------------------------------------------------
// SECURE: parameterized query via execute() + placeholders.
// ---------------------------------------------------------------------------
async function searchOrdersSecure(conn, userId, customerName) {
  // `execute()` uses prepared statements; `?` placeholders are bound
  // as literal values by the driver. The attacker's SQL metacharacters
  // are treated as bytes inside a string, never as SQL syntax.
  //
  // Also: `user_id` is NOT taken from the query string or body. It is
  // taken from the authenticated session. User-supplied IDs in
  // authorization-bearing fields is the other half of this bug class.
  const [rows] = await conn.execute(
    `SELECT id, customer_name, total_cents, created_at
       FROM orders
      WHERE user_id = ?
        AND customer_name = ?`,
    [userId, customerName]
  );
  return rows;
}

// ---------------------------------------------------------------------------
// PREFERRED: use a query builder / ORM that parameterizes by default.
// ---------------------------------------------------------------------------
// Knex:
//   return knex("orders")
//     .where({ user_id: userId, customer_name: customerName })
//     .select("id", "customer_name", "total_cents", "created_at");
//
// Prisma:
//   return prisma.order.findMany({
//     where: { userId, customerName },
//   });
//
// These builders make the parameterization a property of the type
// system. A reviewer sees `.where({ ... })` and can move on; seeing
// `.raw(...)` is the signal to stop and examine the bindings.

// ---------------------------------------------------------------------------
// Demo harness — runs against a disposable MySQL instance. Adjust the
// connection params or delete the block if you do not want to wire it
// up. The point of the demo is to make the failure mode visible.
// ---------------------------------------------------------------------------
async function demo() {
  const conn = await mysql.createConnection({
    host: "127.0.0.1",
    user: "root",
    password: process.env.MYSQL_PWD || "",
    multipleStatements: false,
  });
  await conn.query("CREATE DATABASE IF NOT EXISTS sqli_demo");
  await conn.query("USE sqli_demo");
  await conn.query("DROP TABLE IF EXISTS orders");
  await conn.query(`
    CREATE TABLE orders (
      id INT PRIMARY KEY AUTO_INCREMENT,
      user_id INT NOT NULL,
      customer_name VARCHAR(255) NOT NULL,
      total_cents INT NOT NULL,
      created_at DATETIME DEFAULT CURRENT_TIMESTAMP
    )
  `);
  await conn.query(
    "INSERT INTO orders (user_id, customer_name, total_cents) VALUES (1,'Alice',5000),(1,'Bob',2500),(2,'Eve',9999)"
  );

  const payload = "' OR '1'='1";

  console.log("VULNERABLE:", await searchOrdersVulnerable(conn, 1, payload));
  // → returns all three rows, including user 2's order.

  console.log("SECURE:    ", await searchOrdersSecure(conn, 1, payload));
  // → returns []

  await conn.end();
}

if (require.main === module) {
  demo().catch((err) => { console.error(err); process.exit(1); });
}

module.exports = { searchOrdersVulnerable, searchOrdersSecure };

File: secure-code-review/vulnerable-examples/python/injection/sql_injection.py
"""
SQL Injection — vulnerable and secure variants, side by side.

Scenario: a user-lookup function used by an admin tool. The function
accepts an email (authenticated admin input, but still user-controlled)
and returns the matching user record.

Run it locally against a throwaway SQLite DB to see the exploit hit:

    python sql_injection.py

Both variants are implemented with SQLAlchemy 2.x for parity. The
vulnerable one deliberately uses string interpolation with `text()`,
which is the real-world pattern that leads to incidents — developers
who "know about" parameterized queries still reach for f-strings when
they need "just one dynamic bit" in a raw query.
"""

from sqlalchemy import create_engine, text
from sqlalchemy.orm import sessionmaker, DeclarativeBase, Mapped, mapped_column


class Base(DeclarativeBase): ...


class User(Base):
    __tablename__ = "users"
    id: Mapped[int] = mapped_column(primary_key=True)
    email: Mapped[str]
    role: Mapped[str]


# ---------------------------------------------------------------------------
# VULNERABLE: string interpolation into a text() query.
# ---------------------------------------------------------------------------
def get_user_vulnerable(session, email: str):
    """
    Exploit:
        get_user_vulnerable(session, "nobody@x' OR '1'='1")

    The query becomes:
        SELECT id, email, role FROM users WHERE email = 'nobody@x' OR '1'='1'

    Returns the first row in the table — often the earliest-created admin
    account. On a system where this function is part of a "look up user
    and impersonate" admin tool, this is full account takeover.

    A Union-based variant lets the attacker read arbitrary columns from
    other tables:
        "x' UNION SELECT id, password_hash, role FROM users --"
    """
    query = f"SELECT id, email, role FROM users WHERE email = '{email}'"
    return session.execute(text(query)).first()


# ---------------------------------------------------------------------------
# SECURE: parameterized query via SQLAlchemy bindparams.
# ---------------------------------------------------------------------------
def get_user_secure(session, email: str):
    """
    Same behavior for valid input. For the attacker's payload
    "nobody@x' OR '1'='1", the driver binds the entire string as a
    literal, looks for a user with exactly that email, and returns
    None. No injection is possible.

    Key properties:
      - `:email` is a placeholder, not a format specifier.
      - The value is passed as a dict so the driver can type-bind it.
      - Even if `email` contains SQL, it is treated as a string value.
    """
    return session.execute(
        text("SELECT id, email, role FROM users WHERE email = :email"),
        {"email": email},
    ).first()


# ---------------------------------------------------------------------------
# PREFERRED: use the ORM, which parameterizes by construction.
# ---------------------------------------------------------------------------
def get_user_orm(session, email: str):
    """
    The ORM-first version. Every filter expression the ORM builds is
    parameterized. You never write the SQL; you never write the risk.

    When reviewers see `session.query(...).filter(...)` they can move
    on. When they see `text()`, they stop and look for bindparams.
    """
    return session.query(User).filter(User.email == email).one_or_none()


# ---------------------------------------------------------------------------
# Demo harness — run this file to see the vulnerable version exploited.
# ---------------------------------------------------------------------------
def _demo():
    engine = create_engine("sqlite:///:memory:")
    Base.metadata.create_all(engine)
    Session = sessionmaker(bind=engine)

    with Session() as s:
        s.add_all([
            User(email="admin@example.com", role="admin"),
            User(email="alice@example.com", role="user"),
            User(email="bob@example.com",   role="user"),
        ])
        s.commit()

        payload = "nobody@x' OR '1'='1"

        vuln_result = get_user_vulnerable(s, payload)
        print("VULNERABLE:", vuln_result)
        # → (1, 'admin@example.com', 'admin')

        secure_result = get_user_secure(s, payload)
        print("SECURE:    ", secure_result)
        # → None

        orm_result = get_user_orm(s, payload)
        print("ORM:       ", orm_result)
        # → None


if __name__ == "__main__":
    _demo()
